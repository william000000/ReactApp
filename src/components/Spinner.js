import React from 'react';
import Loader from 'react-loader-spinner';
import '../style.css';

const Spinner = () => (
    <div className="overlay">
        <div className="loader">
            <Loader type="TailSpin" color="#00B9F2" height={100} width={100} />
        </div>
    </div>

);
export default Spinner;
