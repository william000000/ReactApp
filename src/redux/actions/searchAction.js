import actionsTypes from '../actions-types';
import axios from 'axios';
import { notifyError } from '../../helper/notification';
import dotenv from 'dotenv';
dotenv.config();


export const errorHandle = (data) => ({
    type: actionsTypes.ERROR_HANDLER,
    payload: data
});
export const spinnerLoader = (data) => ({
    type: actionsTypes.SPINNER,
    payload: data
});
export const searchByCity = payload => async dispatch => {
    try {
        dispatch(spinnerLoader(true));
        const result = await axios(`${process.env.FETCH_DATA_URL}/${payload}`, {
            method: "GET",
            headers: {
                "x-rapidapi-host": `${process.env.FETCH_DATA_HOST}`,
                "x-rapidapi-key": `${process.env.FETCH_DATA_KEY}`
            }
        });

        dispatch({ type: actionsTypes.FETCHDATA, payload: result.data[0] });
        dispatch(spinnerLoader(false));
    } catch (err) {
        dispatch(errorHandle(err.message));
        notifyError(err.message);
        dispatch(spinnerLoader(false));
    }
};

