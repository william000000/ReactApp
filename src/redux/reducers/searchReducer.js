import initialState from '../store/initialState';
import types from '../actions-types';

const searchReducer = (state = initialState, {
    type,
    payload
}) => {
    switch (type) {
        case types.SAVE_INPUT:
            return {
                ...state,
                input: payload
            };

        case types.FETCHDATA:
            return {
                ...state,
                countryInfo: payload
            };
        case types.ERROR_HANDLER:
            return {
                ...state,
                error: payload
            };
        case types.SPINNER:
            return {
                ...state,
                spinnerLoader: payload
            };
        default:
            return state;
    }
};

export default searchReducer;
