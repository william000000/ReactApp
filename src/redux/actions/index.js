import { searchByCity } from './searchAction';
import { saveInput } from './saveInputAction';

export { searchByCity, saveInput };
