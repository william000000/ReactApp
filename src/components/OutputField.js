import React, { Component } from "react";
import { connect } from 'react-redux';
import { searchByCity } from '../redux/actions';

export class OutputField extends Component {
    render () {
        return (
            <div className="container-output">
                <h2>City Details</h2>
                <div className="outputField">
                    <ul>
                        <li><label>Country Name:</label> {this.props.searchReducer.countryInfo.name}</li>
                        <li><label>Timezone:</label> {this.props.searchReducer.countryInfo.timezones}</li>
                        <li><label>Area:</label> {this.props.searchReducer.countryInfo.area}</li>
                        <li><label>Population: </label>{this.props.searchReducer.countryInfo.population}</li>
                        <li><label>Calling code:</label> {this.props.searchReducer.countryInfo.callingCodes}</li>
                        <li><label>Demonym:</label> {this.props.searchReducer.countryInfo.timezones}</li>
                        <li><label>Region: </label>{this.props.searchReducer.countryInfo.region}</li>
                    </ul>
                </div>
            </div>
        );
    }
}
const mapStateToProps = state => {
    return state;
};


export default connect(mapStateToProps, { searchByCity })(OutputField);
