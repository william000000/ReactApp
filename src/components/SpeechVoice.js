import React, { Component } from "react";
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCoffee, faVolumeUp } from '@fortawesome/free-solid-svg-icons';
import { saveInput } from '../redux/actions';
import img from '../images/icon.png';
import textToSpeech from '../redux/actions/spellAction';


class Speech extends Component {
    constructor (props) {
        super(props);
    }

     handleChange = (e) => {
         e.preventDefault();
         this.props.saveInput(e.target.value);
     }
     render () {
         return (
             <div className="speech-container">
                 <button onClick={() => textToSpeech(this.props.searchReducer.input)}>
                     <FontAwesomeIcon icon={faVolumeUp} className="myIcon"/>
                 </button>
             </div>
         );
     }
}
const mapStateToProps = state => {
    return state;
};

export default connect(mapStateToProps, { saveInput })(Speech);
