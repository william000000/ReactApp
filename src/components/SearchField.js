import React, { Component } from "react";
import { connect } from 'react-redux';
import { ToastContainer } from 'react-toastify';
import { searchByCity, saveInput } from '../redux/actions';
import '../style.css';
import Spinner from './Spinner';
import OutputField from './OutputField';
import SpeechVoice from './SpeechVoice';

export class SearchField extends Component {
    constructor (props) {
        super(props);
    }
    componentDidMount () {
        this.props.searchByCity('kigali');
    }
     handleChange = (e) => {
         e.preventDefault();
         this.props.saveInput(e.target.value);
     }
     render () {
         const { searchByCity, searchReducer } = this.props;

         return (
             <>
                 { searchReducer.spinnerLoader ? (<Spinner />) : (
                     ''
                 ) }
                 <div className="search-container">
                     <h1>City Info</h1>
                     <div className="inputField">
                         <input
                             type="text"
                             placeholder="Search by city name"
                             onChange= {this.handleChange}
                         />
                         <button onClick={() => searchByCity(searchReducer.input)}>Search</button>
                     </div>
                 </div>
                 <SpeechVoice />
                 <OutputField />
                 <ToastContainer/>
             </>
         );
     }
}
const mapStateToProps = state => {
    return state;
};

export default connect(mapStateToProps, { searchByCity, saveInput })(SearchField);
