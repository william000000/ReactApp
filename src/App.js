import React from "react";
import SearchField from "./components/SearchField";
import { Route, Switch, BrowserRouter as Router } from 'react-router-dom';
import "./style.css";

const App = () => {
    return (
        <Router>
            <Switch>
                <Route exact path="/" component={SearchField}/>
            </Switch>
        </Router>
    );
};
export default App;
