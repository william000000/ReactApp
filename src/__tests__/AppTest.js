import React from 'react';
import { shallow, mount } from 'enzyme';
import App from '../App';
import moxios from 'moxios';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { applyMiddleware, createStore } from 'redux';
import configureMockStore from 'redux-mock-store';
import { searchByCity, spinnerLoader, errorHandle } from '../redux/actions/searchAction';
import searchUrl from '../__mocks__/axios';
import OutputField from '../components/OutputField';
import SearchField from '../components/SearchField';
import Spinner from '../components/Spinner';
import SpeechVoice from '../components/SpeechVoice';
import searchReducer from '../redux/reducers/searchReducer';
import combineReducer from '../redux/reducers/index';


let store;
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const mockData = { searchReducer: {
    countryInfo: {
        name: 'Rwanda',
        topLevelDomain: [
            '.rw'
        ],
        alpha2Code: 'RW',
        alpha3Code: 'RWA',
        callingCodes: [
            '250'
        ],
        capital: 'Kigali',
        altSpellings: [
            'RW',
            'Republic of Rwanda',
            'Repubulika y\'u Rwanda',
            'République du Rwanda'
        ],
        region: 'Africa',
        subregion: 'Eastern Africa',
        population: 11262564,
        latlng: [
            -2,
            30
        ],
        demonym: 'Rwandan',
        area: 26338,
        gini: 50.8,
        timezones: [
            'UTC+02:00'
        ],
        borders: [
            'BDI',
            'COD',
            'TZA',
            'UGA'
        ],
        nativeName: 'Rwanda',
        numericCode: '646',
        currencies: [
            'RWF'
        ],
        languages: [
            'rw',
            'en',
            'fr'
        ],
        translations: {
            de: 'Ruanda',
            es: 'Ruanda',
            fr: 'Rwanda',
            ja: 'ルワンダ',
            it: 'Ruanda'
        },
        relevance: '0'
    },
    input: null,
    error: null,
    spinnerLoader: false
} };

const testStore = state => {
    const createStoreWithMiddleware = applyMiddleware(...middlewares)(
        createStore
    );
    return createStoreWithMiddleware(combineReducer, state);
};
store = testStore(mockData);
describe('Test with Enzyme', () => {
    test('Renders', () => {
        expect(shallow(<App />)).toMatchSnapshot();
    });
});

describe('Search Action', () => {
    beforeEach(() => {
        moxios.install(searchUrl);
    });

    afterEach(() => {
        moxios.uninstall(searchUrl);
    });

    it('Should dispatch search successfully', async () => {
        moxios.wait(() => {
            const request = moxios.requests.mostRecent();
            request.respondWith({
                status: 200,
                response: [{
                    name: 'Kigali'
                }]
            });
        });
        store = mockStore({});
        await store.dispatch(searchByCity('kigali')).then(async () => {
            const getActions = store.getActions();
            expect(getActions).toEqual(
                [
                    {
                        payload: true,
                        type: "SPINNER"
                    },
                    {
                        payload: {
                            name: "Kigali"
                        },
                        type: "FETCHDATA"
                    },
                    {
                        payload: false,
                        type: "SPINNER"
                    }
                ]
            );
        });
    });
});

describe('Expect to view output', () => {
    const wrapper = mount(
        <Provider store={store}>
            <OutputField/>
        </Provider>

    );
    it('should view all city info successfully', () => {
        expect(wrapper).toMatchSnapshot();
    });
});
describe('Expect to search', () => {
    const wrapper = mount(
        <Provider store={store}>
            <SearchField/>
        </Provider>

    );
    it('should search successfully', () => {
        expect(wrapper).toMatchSnapshot();
    });
});
describe('Expect to spell the word', () => {
    const wrapper = mount(
        <Provider store={store}>
            <SpeechVoice/>
        </Provider>

    );
    it('should  spell the word successfully', () => {
        expect(wrapper).toMatchSnapshot();
    });
});

describe('Should test the spinner', () => {
    test('renders without crashing', () => {
        expect(shallow(<Spinner />)).toMatchSnapshot();
    });
});

describe('Test reducer', () => {
    it('Should mock fetch data successfully', () => {
        const action = {
            type: "FETCHDATA",
            payload: [mockData.countryInfo]
        };
        const initialState = {
            countryInfo: mockData.searchReducer,
            input: null,
            error: null,
            spinnerLoader: false
        };
        const state = initialState;
        const searchData = searchReducer(state, action);

        expect(searchData).toEqual({ countryInfo: [undefined], error: null, input: null, spinnerLoader: false });
    });

    it('Should mock spinner', () => {
        const action = {
            type: "SPINNER",
            payload: [mockData.countryInfo]
        };
        const initialState = {
            countryInfo: [],
            input: null,
            error: null,
            spinnerLoader: false
        };
        const state = initialState;
        const searchData = searchReducer(state, action);

        expect(searchData).toEqual({
            countryInfo: [],
            input: null,
            error: null,
            spinnerLoader: [undefined]
        });
    });

    it('Should mock error handler', () => {
        const action = {
            type: "ERROR_HANDLER",
            payload: [mockData.countryInfo]
        };
        const initialState = {
            countryInfo: [],
            input: null,
            error: null,
            spinnerLoader: false
        };
        const state = initialState;
        const searchData = searchReducer(state, action);
        expect(searchData).toEqual({
            countryInfo: [],
            input: null,
            error: [undefined],
            spinnerLoader: false
        });
    });
    it('Should mock error handler', () => {
        const action = {
            type: "SAVE_INPUT",
            payload: [mockData.countryInfo]
        };
        const initialState = {
            countryInfo: [],
            input: null,
            error: null,
            spinnerLoader: false
        };
        const state = initialState;
        const searchData = searchReducer(state, action);

        expect(searchData).toEqual({
            countryInfo: [],
            input: [undefined],
            error: null,
            spinnerLoader: false
        });
    });
});

