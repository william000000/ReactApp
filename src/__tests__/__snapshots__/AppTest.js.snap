// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect to search should search successfully 1`] = `ReactWrapper {}`;

exports[`Expect to spell the word should  spell the word successfully 1`] = `ReactWrapper {}`;

exports[`Expect to view output should view all city info successfully 1`] = `ReactWrapper {}`;

exports[`Should test the spinner renders without crashing 1`] = `ShallowWrapper {}`;

exports[`Test with Enzyme Renders 1`] = `ShallowWrapper {}`;
